arngll/src/frame_info.rs

```diff
@@ -0,0 +1,664 @@
+    }
+}
+
+impl Debug for NetworkId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[{:04X}]", self.0)
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum FrameType {
+    Beacon,
+    Data,
+    Ack,
+    MacCommand,
+}
+
+impl FrameType {
+    pub fn try_from_u8(x: u8) -> Option<FrameType> {
+        match x {
+            0 => Some(Self::Beacon),
+            1 => Some(Self::Data),
+            2 => Some(Self::Ack),
+            3 => Some(Self::MacCommand),
+            _ => None,
+        }
+    }
+
+    pub fn to_u8(&self) -> u8 {
+        match self {
+            Self::Beacon => 0,
+            Self::Data => 1,
+            Self::Ack => 2,
+            Self::MacCommand => 3,
+        }
+    }
+}
+
+impl TryFrom<u8> for FrameType {
+    type Error = anyhow::Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        FrameType::try_from_u8(value).ok_or(format_err!("{} is not a valid frame type", value))
+    }
+}
+
+/// Enum encoding the length of a Message Integrity Code.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum MicLen {
+    Mic32 = 0,
+    Mic64 = 1,
+    Mic96 = 2,
+    Mic128 = 3,
+}
+
+impl MicLen {
+    pub fn try_from_u8(x: u8) -> Option<MicLen> {
+        match x {
+            0 => Some(Self::Mic32),
+            1 => Some(Self::Mic64),
+            2 => Some(Self::Mic96),
+            3 => Some(Self::Mic128),
+            _ => None,
+        }
+    }
+
+    pub fn to_u8(&self) -> u8 {
+        match self {
+            Self::Mic32 => 0,
+            Self::Mic64 => 1,
+            Self::Mic96 => 2,
+            Self::Mic128 => 3,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            Self::Mic32 => 4,
+            Self::Mic64 => 8,
+            Self::Mic96 => 12,
+            Self::Mic128 => 16,
+        }
+    }
+}
+
+impl TryFrom<u8> for MicLen {
+    type Error = anyhow::Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        MicLen::try_from_u8(value).ok_or(format_err!("{} is not a valid MIC length", value))
+    }
+}
+
+/// Message Integrity Code.
+#[derive(Clone, Eq, PartialEq)]
+pub struct Mic {
+    pub len: MicLen,
+    pub code: [u8; 16],
+}
+
+impl Debug for Mic {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[{}]", hex::encode(self.as_slice()))
+    }
+}
+
+impl Mic {
+    const EMPTY: Mic = Mic { len: MicLen::Mic32, code: [0u8; 16]};
+
+    pub fn len(&self) -> usize {
+        self.len.len()
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        &self.code[..self.len()]
+    }
+
+    pub fn try_from_slice(slice: &[u8]) -> Result<Mic, anyhow::Error> {
+        if slice.len() < 4 {
+            bail!("Bad MIC size");
+        }
+        let mic_len = MicLen::try_from_u8(((slice.len()/4)-1).try_into()?)
+            .ok_or(format_err!("Bad MIC size"))?;
+        let mut code = [0u8; 16];
+        for i in 0..mic_len.len() {
```

Comment:
Instead of an open-coded loop, perhaps `copy_from_slice`? You could say, perhaps,

    let mut code = [0u8; 16];
    (&mut code[..slice.len()]).copy_from_slice(slice);
    Ok(Mic { len: mic_len, code })

Reply:
Done.